### ISSUE TITLE
Virtualmin fails to detect problems in zones causing unresolvable subdomains.

### ISSUE THREAD

**Issue by chris001** [webmin/webmin#521]
This maybe a virtualmin or webmin issue or both.

It's too easy to accidentally get a subdomain into a state such that it is unresolvable ie DNS returns the error code `NXDOMAIN` (Non eXistent Domain name).  Try to ping the subdomain, open the web site of the subdomain, or do anything with the subdomain, and it's as if the subdomain does not actually exist.
You get back the error, from `dnssec-debugger.verisignlabs.com`:
```
Zone gdomain.com (2001:470:300::2) returns NXDOMAIN for s.gdomain.com
```
This is a Very Bad(TM) Critical bug, because your DNS tells people terribly false information, that your website/subdomain does not exist, when it does, it's just that **virtualmin is failing to detect a semantic error** in the DNS zone for the subdomain.

Workaround appears to be, you have to manually add nameservers for the subdomain, into the parent ie super domain !!

Virtualmin / Webmin should detect this type of error (either the NXDOMAIN error itself, or by linting the zones and detecting a missing record that is logically needed to prevent the sure occurrence of NXDOMAIN), and either alert you and tell you how to fix it, offer to fix it for you, or just fix it without asking, and tell you it fixed it, why, and how.

**Comment by jcameron** [comment]
Is the problem that Virtualmin didn't add the NS records to the parent domain, or that a domain owner can remove them without it causing validation to fail?

**Comment by chris001** [comment]
This is how it happened, IIRC...
1. First, a simple A record or possibly CNAME alias, for one subdomain host was added manually in the virtualmin top level virtual server.  
2. Then, a virtualmin subserver was created for this subdomain, so that a website could be hosted in it, and the website managed by virtualmin.  
3. Various combinations of ipv6 addresses may have been manually updated and added to various records of the subserver, ie MX, AAAA.  
4. Records for the subdomain are then removed from the parent domain (A, CNAME).
5. You get an NXDOMAIN error on any dns lookup for the subdomain.
6. It appears the reason for this is NS records for the subdomain are missing from the parent domain.

**Comment by jcameron** [comment]
Is an NS record needed if both the parent and sub-domains are hosted on the same system and are in the same zone file? I'd have expected that the records would still be resolvable in that case.

**Comment by chris001** [comment]
It appears that, yes, NS record is needed in the parent domain, even when parent and subdomain are hosted on the same bind9 server, in order to avoid the dreaded NXDOMAIN error when trying to resolve the subdomain.  I'm going from experience using the verisign DNS checker tools online, and command line ping/dig/nslookup, on generic windows client and linux server machines.  The DNS RFCs may say something slightly different, or make certain exceptions in certain situations, they tend to be a bit complex due to how DNS has evolved over time since the 80s and arpanet using uucp copies of the `/etc/hosts` file.

**Comment by chris001** [comment]
We should probably try to reproduce it independently, and double triple check it on live running servers.

**Comment by chris001** [comment]
This bug still exists.  The notification on the main screen is very helpful and shows expired DNSSEC signatures, and therefore unresolveable domains, yet it doesn't automatically fix the issue by re-signing the domains or doing whatever it needs to do to get the expired DNS record signatures for the domains, to comply with DNSSEC expiration and re-signing rules.  It might be best to have BIND9 do the work behind the scenes because it knows the rules the be t and as a daemon itself it has the feature to maintain and update the signatures in time so they don't expire and remain expired and become totally unable to resolve and connect to, as is happening now.
```
 Error!
Warning : The following zones have expired DNSSEC signatures : 
server4.edomain.com (14/Feb/2017), p.gdomain.com (19/Apr/2017), bdomain.com (15/Apr/2017), 
pm.gdomain.com (19/Apr/2017), cdomain.com (19/Apr/2017), ipv6.edomain.com (26/Jun/2017), 
gdomain.com (19/Apr/2017), cdomain.com (19/Apr/2017), wdomain.com (03/Jun/2017), 
server2.edomain.com (14/Feb/2017), sdomain.com (18/Apr/2017), bdomain.com (18/Feb/2017), 
rdomain.com (12/Aug/2017), server1.edomain.com (29/Nov/2017), server3.edomain.com (14/Feb/2017), 
new.gdomain.com (19/Apr/2017), idomain.com (18/Feb/2017), shop.gdomain.com (19/Apr/2017), 
cdomain2.com (18/Feb/2017) 
```

**Comment by jcameron** [comment]
Have you enabled scheduled DNSSEC zone re-signing in the BIND DNS Server module?

**Comment by chris001** [comment]
Yes the re-signing is enabled.
I checked the `crontab` and it shows:
```
0 16 * * * /etc/webmin/bind8/resign.pl --debug
```
Does the `--debug` option not actually sign the zone keys.  Does it  only print debug info.

**Comment by jcameron** [comment]
The `--debug` flag just causes it to print debugging information.

You should try running that manually and posting the output here.

**Comment by chris001** [comment]
Running it manually signed the zones. 
Had to run it twice, the first time it signed all but complained about one was missing a signature record. The second time running it, it didn't complain, guess it generated that record and got thru that zone without any problem.   These zones are now all signed with valid active signature records.
Edited the crontab file and removed the `--debug` flag.

**Comment by jcameron** [comment]
Ok - odd that it only worked the second time though!

**Comment by chris001** [comment]
Yeah! I think the first time thru, there was a circular reference it couldn't resolve on the fly, but it updated one record to current and broke the circular ref. The 2nd time thru, the circular reference wasn't there anymore, and it could update that previously impossible to update record.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
